ensime_shared/ensime.py

```diff
@@ -80,44 +91,89 @@ def teardown(self):
         """Say goodbye..."""
         for c in self.clients.values():
             c.teardown()
+        for server in self.servers.values():
+            server.stop()
+
+    def current_client(self):
+        """Get the client for current file in the editor.
+
+        Returns:
+            Optional[EnsimeClient]
+        """
+        config = self.current_project_config()
+        if config:
+            return self.client_for(config.filepath)
 
-    def current_client(self, quiet, bootstrap_server, create_client):
-        """Return the client for current file in the editor."""
+    def current_project_config(self):
+        """Get the project configuration for the current file in the editor.
+
+        Returns:
+            Optional[ProjectConfig]
+        """
         current_file = self._vim.current.buffer.name
         config_path = ProjectConfig.find_from(current_file)
         if config_path:
-            return self.client_for(
-                config_path,
-                quiet=quiet,
-                bootstrap_server=bootstrap_server,
-                create_client=create_client)
-
-    def client_for(self, config_path, quiet=False, bootstrap_server=False,
-                   create_client=False):
+            return ProjectConfig(config_path)
+
+    def client_for(self, config_path):
         """Get a cached client for a project, otherwise create one."""
-        client = None
-        abs_path = os.path.abspath(config_path)
-        if abs_path in self.clients:
-            client = self.clients[abs_path]
-        elif create_client:
-            client = self.create_client(config_path)
-            if client.setup(quiet=quiet, bootstrap_server=bootstrap_server):
-                self.clients[abs_path] = client
-        return client
+        key = os.path.realpath(config_path)
+        if key in self.clients:
+            return self.clients[key]
```

Comment:
Do you think not using `return` may improve readability? I usually avoid several `return`s because they break program logic in unexpected ways and they have a big effect on readability in big chunks of code (as readers lose the thread of the logic and can be easily missed). Though redundant, I usually prefer this style:
```scala
client = None
if key in self.clients:
  client = self.clients[key]
else:
  client = self.create_client(config_path)
return client
```
  
The previous code snippet is very common in Java code, where the idiom of "only return once" is quite used. I'm not aware of Python devs using it and I don't want to bikeshed --all this refactoring is amazing--, but I'm just curious what you think about it.